docs/csharp/roslyn-sdk/tutorials/how-to-write-csharp-analyzer-code-fix.md

```diff
@@ -0,0 +1,615 @@
+
+[!code-csharp[Don't report diagnostics on multiple declarations if all can't be const](~/samples/csharp/roslyn-sdk/Tutorials/MakeConst/MakeConst/MakeConst.Test/MakeConstUnitTests.cs#MultipleDeclarations "Test to ensure no diagnostics are reported when multiple declarations can't all be made const")]
+
+As before, these tests will pass now. It's time to write the code that checks for these conditions. You'll perform some semantic analysis using the <xref:Microsoft.CodeAnalysis.Diagnostics.SyntaxNodeAnalysisContext>. You use the `context` argument to determine whether the local variable declaration can be made `const`. A <xref:Microsoft.CodeAnalysis.SemanticModel?displayProperty=nameWithType> represents of all semantic information in a single source file. You can earn more in the article that covers [semantic models](../work-with-semantics.md). Add the following code to the `AnalyzeNode` method:
+
+```csharp
+// Ensure that all variables in the local declaration have initializers that
+// are assigned with constant values.
+foreach (var variable in localDeclaration.Declaration.Variables)
+{
+    var initializer = variable.Initializer;
+    if (initializer == null)
+    {
+        return;
+    }
+
+    var constantValue = context.SemanticModel.GetConstantValue(initializer.Value);
+    if (!constantValue.HasValue)
+    {
+        return;
+    }
+}
+```
+
+The first `if` in the preceding code exits as soon one variable declaration that doesn't have an initializer was found. The second `if` returns when a declaration has an initializer that is not a constant.
+
+At this point, your code has found a declaration that is initialized with a constant. The declaration can be made constant provided its value is not modified after it has been initialized. You can illustrate that by creating one more test:
+
+[!code-csharp[Don't report diagnostics when variable is assigned to](~/samples/csharp/roslyn-sdk/Tutorials/MakeConst/MakeConst/MakeConst.Test/MakeConstUnitTests.cs#VariableChangesValue "Test to ensure no diagnostics are reported when the variable declaration is assigned to")]
+
+### Make sure the variable isn't assigned after initialization
+
+You'll use the <xref:Microsoft.CodeAnalysis.SemanticModel?displayProperty=nameWithType> to perform data flow analysis on the local declaration statement. Then, you use the results of this data flow analysis to ensure that none of the local variables are written with a new value anywhere else. Call the <xref:Microsoft.CodeAnalysis.ModelExtensions.GetDeclaredSymbol%2A> extension method to retrieve the <xref:Microsoft.CodeAnalysis.ILocalSymbol> for each variable and checking that it isn't contained with the <xref:Microsoft.CodeAnalysis.DataFlowAnalysis.WrittenOutside%2A?displayProperty=nameWithType> collection of the data flow analysis. Add the following code to `AnalyzeNode` after the closing brace of the `foreach` loop:
```

Comment:
> Call the \<xref:Microsoft.CodeAnalysis.ModelExtensions.GetDeclaredSymbol%2A> extension method

As far as I can tell, the code actually calls [the `CSharpExtensions.GetDeclaredSymbol` extension method](https://docs.microsoft.com/en-us/dotnet/api/microsoft.codeanalysis.csharp.csharpextensions.getdeclaredsymbol?view=roslyn-dotnet#Microsoft_CodeAnalysis_CSharp_CSharpExtensions_GetDeclaredSymbol_Microsoft_CodeAnalysis_SemanticModel_Microsoft_CodeAnalysis_CSharp_Syntax_VariableDeclaratorSyntax_System_Threading_CancellationToken_).